modules/planning/math/piecewise_jerk/piecewise_jerk_problem.cc

```diff
@@ -308,9 +309,20 @@ void PiecewiseJerkProblem::set_ddx_bounds(const double ddx_lower_bound,
 }
 
 void PiecewiseJerkProblem::set_x_ref(const double weight_x_ref,
-                                     std::vector<double> x_ref) {
+                                     const std::vector<double>& x_ref) {
```

Comment:
if x_ref will be called by std::move(), better not define it as const&.

Reply:
Removed std::move. Thank you.

Reply:
I think you misunderstand the meaning , you need to use std::move(x_ref); not remove

Reply:
no need to use std::move for const reference.

Reply:
better continue to use std::move, and don't change the interface. i.e. not update it to const&.
(1)std::move has much better performance for large vector. 
(2) this is a very fundamental math Class, been used for a while. don't update its interface unless necessary.

Reply:
Here 'pass by constant reference' has better performance over 'pass-by-value and move'. Both have to copy once either at interface or at function internal. But 'pass-by-value and move' triggered variable 'x_ref_' deallocation before it's gotten reassigned. 

BTW, interface change doesn't affect client calling, should be fine.

Reply:
probably that's exactly what we want, whoever wrote this function initially, to release memory by using smart pointer. 
what do we achieve here by updating it like this? time complexity? space usage?

" interface change doesn't affect client calling, should be fine." When I did the first round code review on this, I had to manually check all the places this function gets called to make sure nothing is impacted (**for the original version, there's an assumption that, this vector will be "cleared" after this function call**. And this would be gone after your change). What if there are hundreds of places to check?  As I said "this is a fundamental math class working fine for a while =/> don't update the interface **unless necessary**. "should be fine" or even "can pass all test cases" is not a standard to evaluate this, this is a common sense of software engineering when you work on a relatively large project.

Reply:
please reference from all the other functions like set_dx_bounds()... They were all using std::move for better performance.

Reply:
@jmtao 
Please see my inline comments.


probably that's exactly what we want, whoever wrote this function initially, to release memory by using smart pointer.
what do we achieve here by updating it like this? time complexity? space usage?

-- Sure. But we are discussing the performance and I've justified why pass-by-const-reference is better. BTW, it's common sense to pass by reference for large structure with/without move op. 


" interface change doesn't affect client calling, should be fine." When I did the first round code review on this, I had to manually check all the places this function gets called to make sure nothing is impacted 

-- A short answer is `you don't have to inspect all client callings`. Both pass-by-constant-reference or pass-by-value avoid modifying values of parameters outside of function. 

(for the original version, there's an assumption on the caller side that, this vector will be "cleared" after this function call because we used std::move. And this assumption would be gone after your change without the callers being aware of). 

-- This assumption only makes sense when dealing with set_xxx functions for memory copy. But in general, it's not wise to clean memory implicitly by calling an external API from caller side, i.e, doing secret thing in one function, which should be avoided in software design. 

What if there are hundreds of places to check? As I said "this is a fundamental math class working fine for a while => don't update the interface unless NECESSARY. "should be fine" or even "can pass all test cases" is not a standard to evaluate this, this is a common sense of software engineering when you work on a relatively large project.

-- Discussion is good and it helps us to reach the best solution. Once the best known method is reached, we can make some efforts to correct error usages in current projects.

Reply:
> please reference from all the other functions like set_dx_bounds()... They were all using std::move for better performance.

@HongyiSun 

Please see my previous comments. 

Here are examples of similar usages: https://github.com/ApolloAuto/apollo/blob/be8acafb4aebc799ffed0467cce97de3a82fbb04/modules/planning/common/obstacle.cc#L697

https://github.com/ApolloAuto/apollo/blob/be8acafb4aebc799ffed0467cce97de3a82fbb04/modules/planning/math/discretized_points_smoothing/fem_pos_deviation_osqp_interface.h#L42

Reply:
Wow, really helpful discussion!

My two cents:
There are two moves here. The caller -> The function parameter -> The final holder.

Current way, passing by `std::vector<double>` and assign with `std::move`, will force the second move as `steal`, but the first move might be `copy` or `steal`, totally based on the user's expertise.
Pros: Have a chance to make 0 copy.
Cons: Ambiguous on behavior and performance. Have a chance to make 2 copies if the caller and callee are not experienced and careful.

The proposed change, passing by `const std::vector<double>&`, will force the first move as `steal`, but the second move is sure to be a copy.
Pros: It's actually the recommended coding style and best practice. Very clear on API and behavior.
Cons: Sure to make 1 copy.

I would suggest a modern `rvalue` parameter which would guarantee 0 copy:
http://www.rovrov.com/blog/2014/12/19/pass-by-rvalue-reference-or-pass-by-value/

Also CC system team for a look.
@storypku @lfcarol @lx18233184051 @changsh726

Reply:
I agreed it's recommended coding style adopted by us probably from the very beginning. 
my point is , for this particular case, PLEASE don't update the interface of a **fundamental** **MATH** class been in use for a while , unless NECESSARY. 
And if you do, please update them all for consistency.  not just a random one you happen to touch. 

I'm out of here. have fun.

Reply:
I agree the idea to pass by value and keep the std::move inside the function, however, there should be a change on the calling of this function, where `piecewise_jerk_problem.set_x_ref(weight_x_ref_vec, x_ref);` in `piecewise_jerk_path_optimizer.cc` should be change into something like `piecewise_jerk_problem.set_x_ref(weight_x_ref_vec, std::move(x_ref));` And that's the way with not a single copy on the whole process.

Reply:
completely agree!
that's an update I think worth doing.

Reply:
Yeah, if we change the API to `std::vector<double>&& x_ref`, the caller must pass parameter with `std::move`. So the 0-copy practice is well guaranteed.

Reply:
“We can do this refactor in future.” then don't make the change in this PR at this point. Keep all function in sync. 
do it well or not at all.  period

Reply:
Looks like we already reached the best solution, I agree with @xiaoxq to change the API to std::vector<double>&& x_ref. I don't mind following current interface style, but it surely should be refactored enforcing "move" call from caller side in future in case 0 copy is targeted.